website/content/en/docs/best-practices/observability-best-practices.md

```diff
@@ -0,0 +1,255 @@
+---
+title: "Operator Observability Best Practices"
+linkTitle: "Observability Best practices"
+weight: 6
+description: This guide describes the best practices concepts for adding Observability to operators.
+---
+
+## Operator Observability Best Practices
+
+In this document we will outline what operators require in order to meet the "Deep Insights" capability level and provide best practices and examples for creating metrics, recording rules and alerts. It is based on the general guidelines in [Operator Capability Levels](https://sdk.operatorframework.io/docs/overview/operator-capabilities/).
+
+**Note:** For technical documentation of how to add metrics to your operator, please read the [Metrics](https://book.kubebuilder.io/reference/metrics.html) section of the Kubebuilder documentation.
+
+### Deep Insights capability level requirements
+
+1. **Health and Performance metrics** for all of the operator components - Implemented based on the guidelines below.
+2. **Metrics Documentation** - All metrics should have documentation.
+3. **Alerts** for when things are not working as expected for each of the operator's components - Implemented based on the guidelines below.
+4. **Alerts Runbooks** - Each alert MUST include a runbook_url annotation and a runbook.
+5. **Alerts and Metrics Tests** - E2E Testing for metrics and alerts and unit tests for alerts.
+6. **Events** - Custom Resources MUST emit custom events for the operations taking place.
+7. **Metering** - Operator leverages Operator Metering.
+
+   **Note:** Metering isn't mandatory at this point in order to meet the "Deep Insights" capability level requirments.
+
+### Metrics Guidelines
+
+1. Metrics `Help` message should be verbose, since it can be used to create auto generated documentation, Like its done here for example [KubeVirt metrics](https://github.com/kubevirt/kubevirt/blob/main/docs/metrics.md) and generated by [KubeVirt metrics doc generator](https://github.com/kubevirt/kubevirt/blob/main/tools/doc-generator/doc-generator.go).
+
+The `help` message should include the following details:
+- What does this metric measure
+- What is the unit of measurement. See the [Prometheus base units](https://prometheus.io/docs/practices/naming/#base-units) and [Understanding metrics types](https://prometheus.io/docs/tutorials/understanding_metric_types/#types-of-metrics)
+- What does the output mean.
+When creating a new metric or recording rule that reports a resource like a ‘pod’ or a ‘container’ name, please make sure that the `namespace` is included, in order for it to be uniquely identified. 
+**Note:** Usually the ‘namespace’ label is populated via service discovery, but there can be cases where it should be added explicitly, usually this can happen for recording rules.
```

Comment:
Also, see that Kubernetes has a doc about the label's good practices and the namespace is not part of them, see: https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/

Why we would need to have a label with the namespace?

Reply:
I hit this specific issue. Its related specifically for k8s. If you want to identify a container/pod you must have its namespace since the same name of a pod/container can live in more than one namespace. To get a 1:1 match you must have the namespace.

Reply:
@sradco are you suggesting labelling "namespaces" [on metrics](https://prometheus.io/docs/practices/naming/#labels)? (where the user does not necessarily need to add label/annotation of namespace on K8s resources.)

Reply:
@Kavinjsir If a metric includes a pod or other resource that is tied to a namspace and its name be the same name in different namespaces, then the developer should make sure the metric includes the namespace so that the user can identify the correct resource in question.